src/Microsoft.Extensions.OData.Migration/Formatters/Deserialization/ODataMigrationInputFormatter.cs

```diff
@@ -0,0 +1,257 @@
+    using Microsoft.AspNet.OData.Formatter;
+    using Microsoft.AspNet.OData.Formatter.Deserialization;
+    using Microsoft.AspNet.OData.Routing;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Http.Headers;
+    using Microsoft.AspNetCore.Mvc.Formatters;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Net.Http.Headers;
+    using Microsoft.OData;
+    using Microsoft.OData.Edm;
+
+    /// <summary>
+    /// Input formatter that supports V3 conventions in request bodies.
+    /// 
+    /// Most of this code was copied from ODataInputFormatter in AspNetCore; the main difference is that instead of using
+    /// DefaultODataDeserializerProvider, a customized ODataMigrationDeserializerProvider is used.
+    /// </summary>
+    public class ODataMigrationInputFormatter : ODataInputFormatter
+    {
+        /// <summary>
+        /// Initialize the ODataMigrationInputFormatter and specify that it only accepts JSON UTF8/Unicode input
+        /// </summary>
+        /// <param name="payloadKinds">The types of payloads accepted by this input formatter.</param>
+        public ODataMigrationInputFormatter(IEnumerable<ODataPayloadKind> payloadKinds)
+            : base(payloadKinds)
+        {
+            SupportedMediaTypes.Add(MediaTypeHeaderValue.Parse("application/json"));
+            SupportedEncodings.Add(Encoding.UTF8);
+            SupportedEncodings.Add(Encoding.Unicode);
+        }
+
+        /// <summary>
+        /// Determine if incoming request is specifically OData v3; if not, then use the next InputFormatter
+        /// </summary>
+        /// <param name="context">InputFormatterContext</param>
+        /// <returns>True if the incoming request is OData v3, otherwise false.</returns>
+        public override bool CanRead(InputFormatterContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            if (context.HttpContext.Request.Headers.ContainsKey("dataserviceversion") ||
+                context.HttpContext.Request.Headers.ContainsKey("maxdataserviceversion"))
+            {
+                return base.CanRead(context);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// If the request has OData v3 headers in it, then process using V3 deserializer provider.
+        /// Otherwise, process as base class.
+        /// </summary>
+        /// <param name="context">InputFormatter context</param>
+        /// <param name="encoding">Encoding of request body</param>
+        /// <returns>InputFormatterResult</returns>
+        public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
```

Comment:
Use nameof(context) instead of a hard-coded string.  Same with all other hard-coded strings that are really names of things.